templates/spa.php

```diff
@@ -0,0 +1,235 @@
+                    }
+
+                    auth0.handleRedirectCallback().then(redirectResult => {
+                        console.log(redirectResult);
+
+                        auth0.getUser().then(user => {
+                            setAppState(true, user);
+
+                            // Issue a request to our quickstart backend api at /api
+                            callBackendApi(auth0);
+                        });
+                    }).catch(() => {
+                        setAppState(false);
+                        callBackendApi();
+                    });
+                });
+            });
+
+            function setAppState(loggedIn, user) {
+                var name = document.getElementById('name');
+                var note = document.getElementById('note');
+                var state = document.getElementById('state');
+                var login = document.getElementById('login');
+                var logout = document.getElementById('logout');
+
+                if (loggedIn === true) {
+                    state.classList.remove('bg-gray-500');
+                    state.classList.add('bg-green-500');
+                    state.innerText = 'Logged in';
+
+                    login.classList.add('hidden');
+                    logout.classList.remove('hidden');
+
+                    name.classList.remove('text-gray-400');
+                    name.classList.add('text-auth0');
+
+                    note.innerText = 'You are authenticated.';
+
+                    if (user) {
+                        name.innerHTML = '<span class="inline-block relative"><img class="h-12 w-12 rounded-full" src="' + user.picture + '"></span> ' + user.name;
+                        note.innerHTML = 'You successfully authenticated as <span class="font-mono bg-gray-800 p-1 rounded text-sm">' + user.sub + '</span>.';
+                    }
+
+                    return;
+                }
+
+                state.classList.remove('bg-green-500');
+                state.classList.add('bg-gray-500');
+                state.innerText = 'Logged Out';
+
+                note.innerText = 'You are not authenticated.';
+
+                logout.classList.add('hidden');
+                login.classList.remove('hidden');
+            }
+
+            function callBackendApi(auth0) {
+                if (auth0) {
+                    auth0
+                    .getTokenSilently()
+                    .then(accessToken =>
+                        fetch('/api', {
+                            method: 'GET',
+                            headers: {
+                                Authorization: `Bearer ${accessToken}`
+                            }
+                        })
+                    )
+                    .then(response => response.json())
+                    .then(response => {
+                        document.getElementById('backendApiResponse').value = JSON.stringify(response, undefined, 4);
+                    });
```

Comment:
Just a nit-pick really (possibly even personal preference) but this whole block could be extracted into a func that takes an access token, to save duplication with the following unauthenticated `fetch` call that does almost exactly the same thing. Just a suggestion but I'll leave it up to you.

Reply:
Cleaned this up a bit, see what you think in the new commit